go/parquet/metadata/file.go

```diff
@@ -0,0 +1,478 @@
+	if f.IsSetEncryptionAlgorithm() {
+		return parquet.AlgorithmFromThrift(f.GetEncryptionAlgorithm())
+	}
+	return parquet.Algorithm{}
+}
+
+func (f *FileMetaData) initSchema() error {
+	root, err := schema.FromParquet(f.FileMetaData.Schema)
+	if err != nil {
+		return err
+	}
+	f.Schema = schema.NewSchema(root.(*schema.GroupNode))
+	return nil
+}
+
+func (f *FileMetaData) initColumnOrders() {
+	orders := make([]parquet.ColumnOrder, 0, f.Schema.NumColumns())
+	if f.IsSetColumnOrders() {
+		for _, o := range f.GetColumnOrders() {
+			if o.IsSetTYPE_ORDER() {
+				orders = append(orders, parquet.ColumnOrders.TypeDefinedOrder)
+			} else {
+				orders = append(orders, parquet.ColumnOrders.Undefined)
+			}
+		}
+	} else {
+		orders = orders[:f.Schema.NumColumns()]
+		orders[0] = parquet.ColumnOrders.Undefined
+		for i := 1; i < len(orders); i *= 2 {
+			copy(orders[i:], orders[:i])
+		}
+	}
+	f.Schema.UpdateColumnOrders(orders)
+}
+
+// WriterVersion returns the constructed application version from the
+// created by string
+func (f *FileMetaData) WriterVersion() *AppVersion {
+	if f.version == nil {
+		f.version = NewAppVersion(f.GetCreatedBy())
+	}
+	return f.version
+}
+
+// SetFilePath will set the file path into all of the columns in each row group.
+func (f *FileMetaData) SetFilePath(path string) {
+	for _, rg := range f.RowGroups {
+		for _, chunk := range rg.Columns {
+			chunk.FilePath = &path
+		}
+	}
+}
+
+// AppendRowGroups will add all of the rowgroup metadata from other to the
+// current file metadata
+func (f *FileMetaData) AppendRowGroups(other *FileMetaData) error {
+	if !f.Schema.Equals(other.Schema) {
+		return xerrors.New("parquet/FileMetaData: AppendRowGroups requires equal schemas")
+	}
+
+	f.RowGroups = append(f.RowGroups, other.GetRowGroups()...)
+	for _, rg := range other.GetRowGroups() {
+		f.NumRows += rg.NumRows
+	}
+	return nil
+}
+
+// Subset will construct a new FileMetaData object containing only the requested
+// row groups by index
+func (f *FileMetaData) Subset(rowGroups []int) (*FileMetaData, error) {
+	for _, i := range rowGroups {
+		if i < len(f.RowGroups) {
+			continue
+		}
+		return nil, xerrors.Errorf("parquet: this file only has %d row groups, but requested a subset including row group: %d", len(f.RowGroups), i)
+	}
+
+	out := &FileMetaData{
+		&format.FileMetaData{
+			Schema:                   f.FileMetaData.Schema,
+			CreatedBy:                f.CreatedBy,
+			ColumnOrders:             f.GetColumnOrders(),
+			EncryptionAlgorithm:      f.FileMetaData.EncryptionAlgorithm,
+			FooterSigningKeyMetadata: f.FooterSigningKeyMetadata,
+			Version:                  f.Version,
+			KeyValueMetadata:         f.KeyValueMetadata(),
+		},
+		f.version,
+		f.Schema,
+		f.FileDecryptor,
+		0,
+	}
+
+	out.RowGroups = make([]*format.RowGroup, 0, len(rowGroups))
+	for _, selected := range rowGroups {
+		out.RowGroups = append(out.RowGroups, f.RowGroups[selected])
```

Comment:
does it pay to make this a map for upstream consumers (I don't think it is done in other implementations just wondering if it is important).

Reply:
a map instead of a slice? Not sure what benefit a map would provide here?

Reply:
it would keep the indices of the RowGroups selected.

Reply:
Ah gotcha. So, in this particular case, it's creating a new `FileMetaData` with the selected subset of RowGroups, I think that it would be surprising to a consumer if the RowGroups in the subset weren't in the order that was provided as the argument (allowing for the rowgroups in the subset to be in a different order than the parent). Thus `RowGroup(0)` of the new FileMetaData that is returned, should be the row group at index `rowGroups[0]` of the parent File Metadata. It might make sense to keep a map *in addition* to the slice, but right now I don't believe it keeps any reference to the parent FileMetaData that it is a subset of, so I'm fine with seeing if that actually gets requested as a feature, rather than doing it now.